pkg/util/ciutil/vars.go

```diff
@@ -34,36 +34,90 @@ type Vars struct {
 	BranchName string
 	// CommitMessage is the full message of the Git commit being built.
 	CommitMessage string
+	// PRNumber is the pull-request ID/number in the source control system.
+	PRNumber string
 }
 
 // DetectVars detects and returns the CI variables for the current environment.
 func DetectVars() Vars {
-	v := Vars{Name: DetectSystem()}
-	// All CI systems have a name (that's how we know we're in a CI system). After detecting one, we will
-	// try to detect some additional CI-specific metadata that the CLI will use. It's okay if we can't,
-	// we'll just have reduced functionality for our various CI integrations.
+	if os.Getenv("PULUMI_DISABLE_CI_DETECTION") != "" {
+		return Vars{Name: ""}
+	}
+
+	systemName := DetectSystem()
+	// If CI variables have been set specifically for Pulumi in the environment,
+	// use that in preference to attempting to automatically detect the CI system.
+	// This allows Pulumi to work with any CI system with appropriate configuration,
+	// rather than requiring explicit support for each one.
+	if os.Getenv("PULUMI_CI_SYSTEM") != "" {
+		v := Vars{}
+		// Override the name with the actual name in the env var.
+		v.Name = System(os.Getenv("PULUMI_CI_SYSTEM"))
+		v.BuildID = os.Getenv("PULUMI_CI_BUILD_ID")
+		v.BuildType = os.Getenv("PULUMI_CI_BUILD_TYPE")
+		v.BuildURL = os.Getenv("PULUMI_CI_BUILD_URL")
+		v.SHA = os.Getenv("PULUMI_CI_PULL_REQUEST_SHA")
+
+		// Don't proceed with automatic CI detection since we are using the PULUMI_* values.
+		return v
+	}
+
+	return detectCISystemVars(systemName)
+}
+
+// detectCISystemVars detects the various aspects of a CI environment
+// using pre-defined variables provided by each CI system.
+// It's okay if we can't detect the additional metadata,
+// we'll just have reduced functionality for our various CI integrations.
+func detectCISystemVars(systemName System) Vars {
+	v := Vars{Name: systemName}
+
 	switch v.Name {
 	case GitLab:
 		// See https://docs.gitlab.com/ee/ci/variables/.
 		v.BuildID = os.Getenv("CI_JOB_ID")
 		v.BuildURL = os.Getenv("CI_JOB_URL")
+		v.BuildType = os.Getenv("CI_PIPELINE_SOURCE")
```

Comment:
Nit: Within each of these per-system blocks, we should use the same field ordering (alphabetical). So `BuildType` should come before `BuildURL`.

Reply:
Done. I actually just realized that I could construct the `BuildURL` for Azure Pipelines as well, but only if the repo is also on Azure DevOps. I'll add that as well.